### PR TITLE
PkgConfigDependency: Move CLI handling into its own abstraction

### PR DESCRIPTION
**Pull request by xclaesse**
This makes the code cleaner and will allow to have other implementations in the future.

### CODE REVIEW

**Comment by tristan957** [comment]
`FAILED run_unittests.py::LinuxlikeTests::test_pkgconfig_gen - AttributeError: 'PkgConfigDependency' object has no attribute 'pkgbin'`

**Comment by tristan957** [comment]
```
mesonbuild/dependencies/scalapack.py:150:18: error: Unexpected keyword argument "define_variable" for "cflags" of "PkgConfigInterface"  [call-arg]
mesonbuild/dependencies/pkgconfig.py:53:5: note: "cflags" of "PkgConfigInterface" defined here
```

**Comment by tristan957** [comment]
```
https://github.com/mesonbuild/meson/actions/runs/5535587348/jobs/10102081828?pr=11973#step:4:1792
```

**Comment by xclaesse** [comment]
@tristan957 got CI to pass now

File: mesonbuild/dependencies/pkgconfig.py
Comment (bruchar1) [inline mesonbuild/dependencies/pkgconfig.py:39]: I think `silent` should be the last argument. That way, you could explicitly write `PkgConfigInterface.instance(enf, for_marchine, silent=False)`, which is self-documenting. Otherwise, we don't know what the bool flag is when the function is invoked.

The same apply for similar functions and constructors (`PkgConfigCLI`, etc.)

File: mesonbuild/dependencies/pkgconfig.py
Comment (bruchar1) [inline mesonbuild/dependencies/pkgconfig.py:54]:
```suggestion
               define_variable: T.Optional[ImmutableListProtocol[str]] = None) -> T.List[str]:
```
You don't need the quotes because annotations are imported from future. Same apply to other functions.

File: mesonbuild/dependencies/pkgconfig.py
Comment (bruchar1) [inline mesonbuild/dependencies/pkgconfig.py:153]:
```suggestion
            if not re.search(f'^{variable_name}$', out, re.MULTILINE):
```

File: mesonbuild/dependencies/pkgconfig.py
Comment (bruchar1) [inline mesonbuild/dependencies/pkgconfig.py:280]:
```suggestion
    def __init__(self, name: str, environment: Environment, kwargs: T.Dict[str, T.Any], language: T.Optional[str] = None) -> None:
```

File: mesonbuild/dependencies/pkgconfig.py
Comment (tristan957) [reply mesonbuild/dependencies/pkgconfig.py:153]: Might as well prefix it with `r` too

**Comment by tristan957** [comment]
@bruchar1 it might make sense before this is merged to rebase your other one to test out the API

File: mesonbuild/dependencies/pkgconfig.py
Comment (bruchar1) [reply mesonbuild/dependencies/pkgconfig.py:153]: > Might as well prefix it with `r` too, assuming that is possible

Why? There are no special characters to escape here....

Comment (tristan957) [reply mesonbuild/dependencies/pkgconfig.py:153]: I just always prefix regex strings with r regardless

**Comment by bruchar1** [comment]
> @bruchar1 it might make sense before this is merged to rebase your other one to test out the API

My plans are rather to wait for this PR before moving forward. Actually, I'm trying to get a cleaner python wrapper that could potentially be merger in pkgconf directly.

File: mesonbuild/dependencies/pkgconfig.py
Comment (xclaesse) [reply mesonbuild/dependencies/pkgconfig.py]: done

File: mesonbuild/dependencies/pkgconfig.py
Comment (xclaesse) [reply mesonbuild/dependencies/pkgconfig.py]: done

File: mesonbuild/dependencies/pkgconfig.py
Comment (xclaesse) [reply mesonbuild/dependencies/pkgconfig.py]: done

**Comment by tristan957** [comment]
test failure

**Comment by xclaesse** [comment]
> test failure

Should be fixed now.

**Comment by tristan957** [comment]
One of the failures look related.

https://github.com/mesonbuild/meson/actions/runs/5741188843/job/15560721128?pr=11973#step:7:1563

What is the `1` in this message?

**Comment by tristan957** [comment]
> I added a commit that implements __str__ on MachineChoice to simplify that case.

Where is that commit?

File: mesonbuild/dependencies/pkgconfig.py
Comment (tristan957) [inline mesonbuild/dependencies/pkgconfig.py:219]: s/run/ran

File: mesonbuild/dependencies/pkgconfig.py
Comment (tristan957) [inline mesonbuild/dependencies/pkgconfig.py:146]: Be consistent with capitalization of `could`

File: mesonbuild/dependencies/pkgconfig.py
Comment (tristan957) [inline mesonbuild/dependencies/pkgconfig.py:154]: You usually hyphenate pkg-config

File: mesonbuild/dependencies/pkgconfig.py
Comment (tristan957) [inline mesonbuild/dependencies/pkgconfig.py:283]: You seem to usually use a lowercase p for pkg-config

File: mesonbuild/dependencies/pkgconfig.py
Comment (tristan957) [inline mesonbuild/dependencies/pkgconfig.py:278]: @eli-schwartz I wonder if we should change this to pkg-config and deprecate pkgconfig similar to your other PR

File: mesonbuild/dependencies/pkgconfig.py
Comment (tristan957) [inline mesonbuild/dependencies/pkgconfig.py:524]: hyphen and capitalization again

File: mesonbuild/dependencies/pkgconfig.py
Comment (xclaesse) [reply mesonbuild/dependencies/pkgconfig.py:215]: I don't think it should be capitalized, elsewhere it has a capital because it's the first word of the sentence. Fixed another place I found it capitalized when not first word of sentence.

File: mesonbuild/dependencies/pkgconfig.py
Comment (xclaesse) [reply mesonbuild/dependencies/pkgconfig.py:146]: done

File: mesonbuild/dependencies/pkgconfig.py
Comment (xclaesse) [reply mesonbuild/dependencies/pkgconfig.py:283]: In this case it is first in the sentence.

File: mesonbuild/dependencies/pkgconfig.py
Comment (xclaesse) [reply mesonbuild/dependencies/pkgconfig.py:278]: We probably should, but I'll leave that to another PR that deals with that specifically.

File: mesonbuild/dependencies/pkgconfig.py
Comment (xclaesse) [reply mesonbuild/dependencies/pkgconfig.py:524]: done

File: mesonbuild/dependencies/pkgconfig.py
Comment (xclaesse) [reply mesonbuild/dependencies/pkgconfig.py:154]: done

File: mesonbuild/dependencies/pkgconfig.py
Comment (xclaesse) [reply mesonbuild/dependencies/pkgconfig.py:219]: done

**Comment by xclaesse** [comment]
> Where is that commit?

I pushed it in the wrong PR lol, fixed.

File: mesonbuild/dependencies/pkgconfig.py
Comment (tristan957) [reply mesonbuild/dependencies/pkgconfig.py:215]: Oh is this not the first word of the sentence?

**Comment by tristan957** [comment]
Seems like this is ready to go in once the CI is in to me if @bruchar1 has no other comments

File: mesonbuild/dependencies/pkgconfig.py
Comment (xclaesse) [reply mesonbuild/dependencies/pkgconfig.py]: "found" is the first word, capitalized it now. Every small details matters I guess... :D

Comment (xclaesse) [reply mesonbuild/dependencies/pkgconfig.py]: Sorry, I misread what you said, I thought you meant the "P" in "pkg-config" lol

File: mesonbuild/dependencies/pkgconfig.py
Comment (bruchar1) [inline mesonbuild/dependencies/pkgconfig.py:161]: nitpick: `out.splitlines()`

**Review by tristan957 — APPROVED** [review]

File: mesonbuild/dependencies/pkgconfig.py
Comment (xclaesse) [reply mesonbuild/dependencies/pkgconfig.py]: done

File: mesonbuild/dependencies/pkgconfig.py
Comment (xclaesse) [reply mesonbuild/dependencies/pkgconfig.py]: done

**Review by bruchar1 — APPROVED** [review]
LGTM

**Pull request merged**
